cpp/src/arrow/ipc/memory.cc

```diff
@@ -145,5 +145,29 @@ Status MemoryMappedSource::Write(int64_t position, const uint8_t* data, int64_t
   return Status::OK();
 }
 
+MockMemorySource::MockMemorySource(int64_t size) : size_(size), pos_(0) {}
+
+Status MockMemorySource::Close() {
+  return Status::OK();
+}
+
+Status MockMemorySource::ReadAt(
+    int64_t position, int64_t nbytes, std::shared_ptr<Buffer>* out) {
+  return Status::OK();
+}
+
+Status MockMemorySource::Write(int64_t position, const uint8_t* data, int64_t nbytes) {
+  pos_ = std::max(pos_, position + nbytes);
```

Comment:
The goal here is to determine how many bytes there are between the beginning of the buffer and the location where the last byte is being written; the function GetRowBatchSize will most of the time be used to determine how much shared memory should be allocated for IPC and then this is the quantity we care about; if memory is noncontiguous, it is not clear what the desired behaviour is.